core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java

```diff
@@ -172,6 +184,21 @@ private boolean addMember(Address joiner, CacheJoinInfo joinInfo) {
             return false;
          }
 
+         // Validate if the member is allowed to join
+         if (persistentState.isPresent()) {
```

Comment:
Ok, if the wrong node happens to start first, the `getCache()` fails but no state/data is lost, and the user can restart the cluster again. I guess it's just a matter of how good our error messages are going to be ;)

I'm a bit more concerned about nodes without state (or just with a different state) becoming coordinator (via merge) while a cluster with state is starting up. E.g. some nodes already sent their join requests to the old coordinator and got the ok to join, but the new coordinator is not going to try and restore that state. I'm not sure how that will play out with the current code, but I think we need some handling of state during cluster recovery, if only to log better error messages.